File: src/algorithm_java/leetcode/Problem85.java
package algorithm.leetcode;

import java.util.Stack;

public class Problem85 {

    public static int largestRectangle(int[] heights) {
        int[] heightsp = new int[heights.length + 1];
        for (int i = 0; i < heights.length; i++) {
            heightsp[i] = heights[i];
        }
        heightsp[heights.length] = 0;
        int maxSquare = 0;
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        for (int i = 1; i < heightsp.length; i++) {
            while (!stack.empty() && heightsp[stack.peek()] > heightsp[i]) {
                int popIndex = stack.pop();
                int width = stack.empty() ? i : (i - stack.peek() - 1);
                int square = width * heightsp[popIndex];
                if (square > maxSquare) {
                    maxSquare = square;
                }
            }
            stack.push(i);
        }
        return maxSquare;
    }

    public static int maximalRectangle(char[][] matrix) {
        int height = matrix.length;
        if (height == 0)
            return 0;
        int width = matrix[0].length;
        int maxSquare = 0;
        for (int i = 0; i < height; i++) {
            int[] h = new int[width];
            for (int j = 0; j < width; j++) {
                for (int k = i; k >= 0 && matrix[k][j] == '1'; k--) {
                    h[j]++;
                }
            }
            int square = largestRectangle(h);
            if (square > maxSquare) {
                maxSquare = square;
            }
        }
        return maxSquare;
    }

    public static void main(String[] args) {
        char[][] matrix = new char[][] {
                {'1','0','1','0','0'},
                {'1','0','1','1','1'},
                {'1','1','1','1','1'},
                {'1','0','0','1','0'}
        };
        System.out.println(maximalRectangle(matrix));
        int[] t = new int[] {3,1,3,2,2};
        System.out.println(largestRectangle(t));
    }

}

File: src/algorithm_cpp/problem85.cpp
#include<iostream>
#include<vector>
#include<stack>

using namespace std;

class Solution {
public:

    int maxHeight(vector<int> heights) {
	heights.insert(heights.begin(), 0);
	heights.push_back(0);
	stack<int> s;
	s.push(0);
	int rt = 0;
	for (int i = 0; i < heights.size(); i++) {
	    while(heights[s.top()] > heights[i]) {
		int height = heights[s.top()];
		s.pop();
		int width = i - s.top() - 1;
		rt = max(rt, height * width);
	    }
	    s.push(i);
	}
	return rt;


    }

    int maximalRectangle(vector<vector<char>>& matrix) {
	if (matrix.size() == 0)
	    return 0;
	int m = matrix.size();
	int n = matrix[0].size();
	vector<vector<int>> heights(m, vector<int>(n, 0));
	for (int i = 0; i < m; i++) {
	    for (int j = 0; j < n; j++) {
		if (matrix[i][j] == '1') {
		    heights[i][j] = i == 0 ? 1 : heights[i - 1][j] + 1;
	    	}
	    }
	}
	int rt = 0;
	for (int i = 0; i < m; i++) {
	    int height = maxHeight(heights[i]);
	    if (height > rt)
		rt = height;
	}
	return rt;
    }
};

int main() {
    Solution s;
    vector<int> input{1};
    cout << s.maxHeight(input);
}
